### DISCUSSION TITLE
Packaging and moving glsp-client/glsp-server into Theia blueprint electron

### DISCUSSION THREAD

**Discussion by amiths89** [eclipse-glsp/glsp#1409] [General] [answered]
I was trying to package up my code currently based on the glsp-examples/java-emf-theia template into an electron app. Is it pretty straightforward to create an electron app from this example, or would it be a bigger change in the structure of the code?

Edit:
Actually, I was hoping i could integrate the glsp examples into the theia blueprint code. It has a lot of things for free, like the splash screen and even the electorn-builder.yml configuration file.

**Comment by martin-fleck-at** [comment]
I don't think there should be any issues since there is already a browser app or is there anything in particular that you are facing?

**Reply by amiths89** [reply]
Well its not exactly error but its quite unclear to me when i try to connect the glsp-client glsp-server architecture into the theia blueprint. 
I am trying to add the tasklist-theia, tasklist-glsp from the glsp-examples code into the theia-extensions folder and also glsp-server into the root folder in the blueprint code. The build is currently failing due to some configuration issues like tsconfig which i am hoping to get around. So i want to make sure am on the right path and am not missing something here. Also if there is a better and easier alternative solution to this please let me know.
<img src="https://github.com/user-attachments/assets/46811ead-165e-4999-b940-dc7f09a2a2c9" width="300" />

**Reply by amiths89** [reply]
The electron app is building, but the glsp is not working. So my confusion at the moment is how do I bind the glsp server to the theia extension, i guess this where i am going wrong. Would I need to merge the tasklist-theia code and backend-contribution code into either the launcher/product extensions?

**Reply by martin-fleck-at** [reply, marked as answer]
So the way the GLSP server is started and connected to Theia is controlled by the `GLSPServerContribution` and for the tasklist that would be the `TaskListGLSPServerContribution` specifically. In it's core this contribution makes sure that the (frontend) client connections are properly forwarded to the GLSP server. There is already a base implementation that supports socket and websocket connections. The general setup is controlled by the `GLSPServerContributionOptions` (and `GLSPSocketServerContributionOptions`). There are two ways to start the GLSP server:
- Either you start the GLSP server manually before starting the Theia application. We call this sometimes the "external" scenario and this may make sense when your GLSP server also servers as a general server for other purposes. So Theia assumes the server is already running. This is reflected by the [launchedExternally](https://github.com/eclipse-glsp/glsp-theia-integration/blob/7841009f67fdb8cc4b45e03a25309388bde1ebbb/packages/theia-integration/src/node/glsp-server-contribution.ts#L63) property of the server contribution options.
- You let Theia start the GLSP server either on demand (widget open) or application start. This is reflected by the [launchOnDemand](https://github.com/eclipse-glsp/glsp-theia-integration/blob/7841009f67fdb8cc4b45e03a25309388bde1ebbb/packages/theia-integration/src/node/glsp-server-contribution.ts#L58) flag. In this case you also need to give the path to the executable that needs to be started. You can see that in the [TaskListGLSPServerContribution](https://github.com/eclipse-glsp/glsp-examples/blob/e55606e37be62888309a97b765a1335807477a13/project-templates/java-emf-theia/glsp-client/tasklist-theia/src/node/glsp-server-contribution.ts#L30) where the path to the JAR file is given.

So could it be that this path is no longer correct after you copied the files around?

**Reply by amiths89** [reply]
Thanks, @martin-fleck-at, for the reply. I made sure the path to the jar was correct. Actually, it's the same as the initial glsp-examples. Since I placed tasklist-theia under theia-extensions and glsp-server at the root level, it is now the equivalent structure of glsp-client and glsp-server being at the same level in the example i.e. theia-extensions being the equivalent of glsp-client, so the path to the jar remains the same

**Reply by martin-fleck-at** [reply]
Did you manage to get it to work? If not, is there a branch or anything I could maybe try out? It seems that you are combining public repositories so maybe there is a way I can see what it is you are doing.

**Reply by amiths89** [reply]
Sorry, I tried to create a new repo, but it kept taking a long time, and the connection broke, and I wasn't able to branch off the Theia Blueprint. I guess I need permission. Although now I have resolved one of my key issues, which was due to the version of eclipse-glsp dependencies by updating all to 2.2.1. Now, when I launch the code from vs code i.e. in debug mode, it works. But when I launch the electron build, the glsp file doesn't start up(example.tasklist). If this continues, i ll try again and set up a new repository. Also, could u tell me if there's anything I need to change in the start script for electron package.json.
There are my current scripts in electron
```"scripts": {
    "clean": "theia clean && rimraf node_modules",
    "clean:dist": "rimraf dist",
    "build": "yarn -s rebuild && theia build --app-target=\"electron\" --mode development",
    "build:prod": "yarn -s rebuild && theia build --app-target=\"electron\"",
    "rebuild": "theia rebuild:electron --cacheRoot .",
    "watch": "concurrently -n compile,build \"theiaext watch --preserveWatchOutput\" \"theia build --watch --mode development\"",
    "start": "electron scripts/theia-electron-main.js --plugins=local-dir:../../plugins ",
    "start:debug": "yarn start --debug --log-level=debug",
    "package": "yarn clean:dist && yarn rebuild && electron-builder -c.mac.identity=null --publish never",
    "package:prod": "yarn deploy",
    "deploy": "yarn clean:dist && yarn rebuild && electron-builder -c.mac.identity=null --publish always",
    "package:preview": "yarn clean:dist && yarn rebuild && electron-builder -c.mac.identity=null --dir",
    "update:checksum": "ts-node scripts/update-checksum.ts",
    "update:blockmap": "ts-node scripts/update-blockmap.ts",
    "update:theia": "ts-node ../../scripts/update-theia-version.ts",
    "update:next": "ts-node ../../scripts/update-theia-version.ts next",
    "test": "mocha --timeout 60000 \"./test/*.spec.js\"",
    "lint": "eslint --ext js,jsx,ts,tsx scripts && eslint --ext js,jsx,ts,tsx test",
    "lint:fix": "eslint --ext js,jsx,ts,tsx scripts --fix && eslint --ext js,jsx,ts,tsx test -fix"
```
i mean i have not changed either the launch.json or the scripts in the package.json file inside applications/electron from the blueprint repo https://github.com/eclipse-theia/theia-blueprint but it seems to work fine in the vscode debug mode somehow.

**Reply by amiths89** [reply]
@martin-fleck-at I see the below log message in the logs when i launch from VS code which i dont see in my electron build log. Is there a way to set the similar policy in my electron build as well or is there a way around this to get the glsp running?

```
VM153 renderer_init:2 Electron Security Warning (Insecure Content-Security-Policy) This renderer process has either no Content Security
  Policy set or a policy with "unsafe-eval" enabled. This exposes users of
  this app to unnecessary security risks.

For more information and help, consult
https://electronjs.org/docs/tutorial/security.
This warning will not show up
once the app is packaged.
warnAboutInsecureCSP @ VM153 renderer_init:2
```

**Reply by martin-fleck-at** [reply]
Very interesting, so you are calling `yarn start` and `yarn start:debug`, is that correct? Because they are essentially the same with just the `--debug` and `--log-level=debug` flag being set and I'd hope that they don't really have an influence on the start of the GLSP server. And there is nothing else in the backend logs or frontend logs (console log of the Developer tools) that would indicate why the GLSP server would not start up? What if you were to add some console logs to the [TaskListGLSPServerContribution](https://github.com/eclipse-glsp/glsp-examples/blob/e55606e37be62888309a97b765a1335807477a13/project-templates/java-emf-theia/glsp-client/tasklist-theia/src/node/glsp-server-contribution.ts#L30) to see if something is throwing an exception somehow?

The warning you see in the log shouldn't cause any major problems, it is probably cause by the CSP of the [TaskListEditorProvider](https://github.com/eclipse-glsp/glsp-examples/blob/e55606e37be62888309a97b765a1335807477a13/project-templates/node-json-vscode/tasklist-vscode/extension/src/tasklist-editor-provider.ts#L52). However, as the log says, this should not show up in the packaged app.

This is really a curious problem and, unfortunately, I've not encountered that before.

**Reply by amiths89** [reply]
Thanks @martin-fleck-at . I looked at the stack trace by running it from command line, and you were right about the jar path. Although it was fine for the project and vscode build. The electron package was different, so I had to update the electron builder yaml file to create a similar glsp-server jar path, and now it works.
